### ISSUE TITLE
Enable choosing the interface identifier on interface creation

### ISSUE THREAD

**Issue by bimbar** [opnsense/core#7522]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [X] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [X] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Is your feature request related to a problem? Please describe.**

There are two scenarios:
- yesterday a customer lost his interfaces in some way I can't quite fathom and then recreated them in the wrong order causing all firewall rules to apply to the wrong interfaces. The only remedy was to grep the correct assignments out of a restored backup and then delete and recreate all interfaces in the right order.
- Also, in HA scenarios it would be easier to create interfaces, especially on differing hardwares.

**Describe the solution you like**

On creating an interface, make it possible to either choose the interface id outright, or choose the number X in optX.

**Describe alternatives you considered**

Creating interfaces in the right order is possible, but especially in the case of non-contiguous numbering, which can happen if an interface is deleted later, this is not very convenient.

**Additional context**

Discussed it with franco: https://forum.opnsense.org/index.php?topic=40985.0;topicseen

**Comment by AdSchellevis** [comment]
I think the easiest option would be to add an optional field in `Interfaces: Assignments` which accepts `lan`,`wan`, `optXXX` and when not provided pick the first free one in the same order.

Which requires changes in 

https://github.com/opnsense/core/blob/df7adc41c2f1c61a466be2b66c390dea206f5252/src/www/interfaces_assign.php#L118-L125

and a validation (check pattern and not used) around here:

https://github.com/opnsense/core/blob/df7adc41c2f1c61a466be2b66c390dea206f5252/src/www/interfaces_assign.php#L113

Plus side would be that when you drop `wan` and recreate without optional argument, it would also recreate `wan` (which is currently not supported).

**Comment by odin81** [comment]
Would be a nice feature. I try to change my setup from single node to ha, bute I'm not able to create the interfaces in the same order on the new node.

**Comment by melroy89** [comment]
I personally would also love to change this, to avoid any confusion mainly. Like the identifier is currently `lan` and `wan`, but if your WAN is on the SFP+ port (opt1), it could be very confusing in general.

Maybe the BSD kernel just assign those identifiers by default? But if that is true, how is lan and wan pre-defined?

**Comment by fichtner** [comment]
I don’t think it’s a good idea to re-add “lan” or “wan” automatically, but if as an optional input it’s ok. “wan” is hardly special and “lan” is only important in a default configuration because it has an allow all traffic rule. Retrofitting this into the interface assignments consistently is tricky because the user either will expect it or not and adding an additional knob for it will be a slippery slope and avoid getting rid of “wan” and “lan” specialties in the long term.

**Comment by fichtner** [comment]
Thinking a bit further we would need to emulate the port assignment console workflow. Essentially adding a dropdown if you want to create a LAN or WAN or OPT device and act accordingly with a number of help paragraphs to explain it. For the opt device allowing to also set the index… and later convert the monster to MVC. 😉

**Comment by fichtner** [comment]
(OPT being the preselected value to retain current behaviour without surprises.)

**Comment by AdSchellevis** [comment]
The main reason for switching internal references is high-availability and/or syncing settings between machines (also the reason why I'm keeping this on the backlog). In all other cases the internal names are usually just that, internal...

**Comment by fichtner** [comment]
I agree. “Switching” is a misleading word since the process requires replacing the wrongly assigned interface due to the fact that we will not be able to cleanly replace all references to the internal identifier for a very long time.

There is opportunity for a number of best practices here like avoiding WAN/LAN in the first place or reusing the same config for the HA backup to start with… The creation of arbitrary OPT definitely does help but most of this also revolves around people noticing only after the fact and asking for a programmatic fix which is a bottomless pit.

**Comment by AdSchellevis** [comment]
There certainly is a high risk of doing too much here, eventually we'll come up with a simple and focused fix, knowing it can't fix everything in the world (nor should it try to).

**Comment by mhalano** [comment]
I just installed my OPNsense, and the LAN interface has `lan` as identifier, WAN has `wan` as identifier but my IOT interface has `opt3`. Not cool.

**Comment by breakone9r** [comment]
Instead of using the interfaces directly, all firewall rules should instead just use groups, named as WAN, LAN, etc,

Keep the hardware names of the interfaces, and then just add/remove them from the LAN or WAN group(s)

This has the added benefit of firewall rules not needing to be updated if you change hardware, you just put the new interface into the proper interface group.

Another benefit to this is allowing us to use multiple interfaces as part of the same lan group.

Setting IP and such should still be at the interface level. Just adding a layer of abstraction at the firewall level, which should also ease creation of HA nodes.
